src/learning/gmm.rs

```diff
@@ -209,174 +212,404 @@ impl GaussianMixtureModel {
         &self.mix_weights
     }
 
+    /// The model mean likelihood
+    ///
+    /// Returns the mean log likelihood of the model
+    pub fn log_lik(&self) -> f64 {
+        self.log_lik
+    }
+
     /// Sets the max number of iterations for the EM algorithm.
     ///
     /// # Examples
     ///
     /// ```
-    /// use rusty_machine::learning::gmm::GaussianMixtureModel;
+    /// use rusty_machine::learning::gmm::{GaussianMixtureModel, Random, CholeskyFull};
     ///
-    /// let mut gmm = GaussianMixtureModel::new(2);
+    /// let mut gmm: GaussianMixtureModel<Random, CholeskyFull> = GaussianMixtureModel::new(2);
     /// gmm.set_max_iters(5);
     /// ```
     pub fn set_max_iters(&mut self, iters: usize) {
         self.max_iters = iters;
     }
 
-    fn initialize_covariances(&self, inputs: &Matrix<f64>, reg_value: f64) -> LearningResult<Matrix<f64>> {
-        match self.cov_option {
-            CovOption::Diagonal => {
-                let variance = try!(inputs.variance(Axes::Row));
-                Ok(Matrix::from_diag(&variance.data()) * reg_value.sqrt())
-            }
+    // == Parameters
+    // inputs : [n_samples, n_features]
+    //
+    // == Returns
+    // weighted_log_prob : [n_features, n_components]
+    //
+    fn estimate_weighted_log_prob(&self, inputs: &Matrix<f64>) -> Matrix<f64> {
+        let mut log_prob = self.estimate_log_prob(&inputs);
+        // println!("log_prob: \n{:.4}", log_prob.select_rows(&[0, 1, 2, 3, 4, 5, 6]));
+        // println!("mix_weights: \n{:?}", &self.mix_weights);
+        let log_weights = self.mix_weights.iter().map(|w| w.ln());
+        for (lp, lw) in log_prob.iter_mut().zip(log_weights) {
+            *lp += lw;
+        }
+        log_prob
+    }
 
-            CovOption::Full | CovOption::Regularized(_) => {
-                let means = inputs.mean(Axes::Row);
-                let mut cov_mat = Matrix::zeros(inputs.cols(), inputs.cols());
-                for (j, row) in cov_mat.iter_rows_mut().enumerate() {
-                    for (k, elem) in row.iter_mut().enumerate() {
-                        *elem = inputs.iter_rows().map(|r| {
-                            (r[j] - means[j]) * (r[k] - means[k])
-                        }).sum::<f64>();
-                    }
+    // called estimate_log_gaussian_prob in scipy
+    //
+    // == Paramers
+    // inputs : [n_samples, n_features]
+    //
+    //
+    fn estimate_log_prob(&self, inputs: &Matrix<f64>) -> Matrix<f64> {
+        let precisions: &Vec<Matrix<f64>> = self.precisions.as_ref().unwrap();
+        let ref model_means = self.model_means.as_ref().unwrap();
+        // The log of the determinant for each precision matrix
+        let log_det = precisions.iter().map(|m| m.det().ln());
+        // println!("log_det: {:?}", log_det);
```

Comment:
Can we remove this commented line please?

_Edit: And the other commented `println!`s below_